System/Metrics/Dimensional.hs

```diff
@@ -0,0 +1,106 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE DeriveDataTypeable #-}
+
+-- | This module defines a type for mutable dimensional string-valued label.
+-- Dimensional label are variable values and can be used to track e.g. the
+-- current number of call to an api endpoint by its http return code.
+-- All operations on Dimensional label are thread-safe.
+module System.Metrics.Dimensional where
+
+import qualified System.Metrics.Counter as Counter
+import qualified System.Metrics.Distribution as Distribution
+import qualified System.Metrics.Gauge as Gauge
+import qualified System.Metrics.Label as Label
+
+import Control.Applicative ((<$>), pure)
+import Control.Exception (throwIO)
+import Data.HashMap.Strict (HashMap, empty)
+import qualified Data.HashMap.Strict as HashMap
+import Data.IORef (IORef, atomicModifyIORef', newIORef, readIORef)
+import Data.Text (Text)
+import Data.Typeable
+import GHC.Exception (Exception)
+import Prelude hiding (lookup)
+
+type Name = Text
+
+type Explanation = Text
+
+type Dimensions = [Text]
```

Comment:
It might be convenient to allow a dimension to be a member of some `Dimension` type class with instances for `Text`, `String`, `Int`, and so forth. Will make the API nicer to use without having to use "to text" conversions at each call site for dimensions that are e.g. naturally ints (e.g. like HTTP status codes). The type class can be used just in the user-facing API. Internally it can all be `Text`.